File: Prashant/searching/BinarySearch.java

/** this program is implementation of binary search */
import java.util.*;

class BinarySearch {
    public static void binarySearch(int arr[], int first, int last, int key) {
        int mid = (first + last) / 2;
        while (first <= last) {
            if (arr[mid] < key) {
                first = mid + 1;
            } else if (arr[mid] == key) {
                System.out.println("Element is found at index: " + mid);
                break;
            } else {
                last = mid - 1;
            }
            mid = (first + last) / 2;
        }
        if (first > last) {
            System.out.println("Element is not found!");
        }
    }

    public static void main(String args[]) {
        int arr[] = { 10, 20, 30, 40, 50 };
        int key = 30;
        int last = arr.length - 1;
        binarySearch(arr, 0, last, key);
    }
}
// Best-case performance O(1)
// Average performance O(log n)
// Worst-case space complexity O(1)

File: RachitGoyal/BinarySearch.cpp
#include<bits/stdc++.h>
using namespace std;

void binarysearch(int n,int arr[],int k)
{
    int beg=0,flag=0;
    int end=n-1;
    
    while(beg<=end)
    {
        int mid=(beg+end)/2;
        if(arr[mid]==k)
        {
            
            flag++;
            cout<<"search suceesfull";
            break;
        }
        else if(arr[mid]>k)
        {
            end=mid-1;
        }
        else
        {
            beg=mid+1;
        }
    }
    if(flag==0)
    {
        cout<<"NOT PRESENT";
    }
    
    
}
int main()
{
    int n,k;
    cin>>n>>k;
    int arr[n];
    for(int i=0;i<n;i++)
    {
        cin>>arr[i];
        
    }
    binarysearch(n,arr,k);
    
    return 0;
}



//Space Complexity: O(1)
//Time Complexity:-
//Worst Case: O(log n)
//Best Case: O(1)
